File: RootTools2/src/main/java/com/rarnu/tools/neo/utils/RootUtils.java
package com.rarnu.tools.neo.utils;

import android.util.Log;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStreamReader;

public class RootUtils {

    public static class CommandResult {
        public String result = "";
        public String error = "";
    }

    interface CommandCallback {
        void onReadLine(String line);

        void onReadError(String line);

        void onCommandFinish();
    }

    public static CommandResult runCommand(String command, boolean root) {
        return runCommand(new String[]{command}, root, null);
    }

    public static CommandResult runCommand(String[] prog) {
        return runCommand(prog, false, null);
    }

    public static CommandResult runCommand(String[] prog, CommandCallback callback) {
        return runCommand(prog, false, callback);
    }

    public static CommandResult runCommand(String command, boolean root, CommandCallback callback) {
        return runCommand(new String[]{command}, root, callback);
    }

    public static CommandResult runCommand(String[] command, boolean root, CommandCallback callback) {
        Log.e("RootUtils", command[0]);
        Process process = null;
        DataOutputStream os = null;
        BufferedReader brOut = null;
        BufferedReader brErr = null;
        CommandResult ret = new CommandResult();
        try {
            StringBuffer output = new StringBuffer();
            StringBuffer error = new StringBuffer();
            if (root) {
                process = Runtime.getRuntime().exec("su");
                os = new DataOutputStream(process.getOutputStream());

                os.writeBytes(command[0] + "\n");
                os.writeBytes("exit\n");
                os.flush();
            } else {
                if (command.length == 1) {
                    process = Runtime.getRuntime().exec(command[0]);
                } else {
                    process = Runtime.getRuntime().exec(command);
                }
            }

            String line;
            brOut = new BufferedReader(new InputStreamReader(process.getInputStream()));
            while ((line = brOut.readLine()) != null) {
                output.append(line).append('\n');
                if (callback != null) {
                    callback.onReadLine(line);
                }
            }
            brErr = new BufferedReader(new InputStreamReader(process.getErrorStream()));
            while ((line = brErr.readLine()) != null) {
                error.append(line).append('\n');
                if (callback != null) {
                    callback.onReadError(line);
                }
            }
            process.waitFor();
            ret.result = output.toString().trim();
            ret.error = error.toString().trim();

            // ignore warning output
            if (ret.error.startsWith("WARNING:")) {
                ret.error = "";
            }
        } catch (Exception e) {
            ret.result = "";
            ret.error = e.getMessage();
        } finally {
            try {
                if (os != null) {
                    os.close();
                }
                if (brOut != null) {
                    brOut.close();
                }
                if (brErr != null) {
                    brErr.close();
                }
            } catch (Exception e) {
                ret.result = "";
                ret.error = e.getMessage();
            }
        }
        if (callback != null) {
            callback.onCommandFinish();
        }
        return ret;
    }

}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/utils/RootUtils.kt
package com.rarnu.tools.neo.utils

import java.io.BufferedReader
import java.io.DataOutputStream
import java.io.InputStreamReader

object RootUtils {

    data class CommandResult(var result: String, var error: String)

    interface CommandCallback {
        fun onStart()
        fun onReadLine(line: String?)
        fun onReadError(line: String?)
        fun onFinish()
    }

    fun runCommand(command: String, root: Boolean): CommandResult = runCommand(arrayOf(command), root, null)

    fun runCommand(prog: Array<String>): CommandResult = runCommand(prog, false, null)

    fun runCommand(prog: Array<String>, listener: CommandCallback?): CommandResult = runCommand(prog, false, listener)

    fun runCommand(command: String, root: Boolean, listener: CommandCallback?): CommandResult = runCommand(arrayOf(command), root, listener)

    fun runCommand(command: Array<String>, root: Boolean, listener: CommandCallback?): CommandResult {
        val ret = CommandResult("", "")
        var process: Process?
        var rootOs: DataOutputStream? = null
        var procOutOs: BufferedReader? = null
        var procErrOs: BufferedReader? = null
        listener?.onStart()
        try {
            if (root) {
                process = Runtime.getRuntime().exec("su")
                rootOs = DataOutputStream(process.outputStream)
                rootOs.writeBytes("${command[0]}\n")
                rootOs.writeBytes("exit\n")
                rootOs.flush()
            } else {
                if (command.size == 1) {
                    process = Runtime.getRuntime().exec(command[0])
                } else {
                    process = Runtime.getRuntime().exec(command)
                }
            }
            procOutOs = BufferedReader(InputStreamReader(process.inputStream))
            procErrOs = BufferedReader(InputStreamReader(process.errorStream))
            var outStr = StringBuffer()
            var errStr = StringBuffer()
            var line: String?
            while (true) {
                line = procOutOs.readLine()
                if (line != null) {
                    outStr.append("${line}\n")
                    listener?.onReadLine(line)
                } else {
                    break
                }
            }
            while (true) {
                line = procErrOs.readLine()
                if (line != null) {
                    errStr.append("${line}\n")
                    listener?.onReadError(line)
                } else {
                    break
                }
            }
            process.waitFor()
            ret.result = outStr.toString().trim { it <= ' ' }
            ret.error = errStr.toString().trim { it <= ' ' }
        } catch(e: Exception) {
            if (e.message != null) {
                ret.error = if (e.message == null) "" else e.message!!
            }
        } finally {
            rootOs?.close()
            procOutOs?.close()
            procErrOs?.close()
        }
        listener?.onFinish()
        return ret
    }

}
